### ISSUE TITLE
use-before-declare: false positive in destructuring assignment

### ISSUE THREAD

**Issue by tamird** [palantir/tslint#2551]
### Bug Report

- __TSLint version__: 
```
yarn run tslint -- --version
yarn run v0.22.0
$ "/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/node_modules/.bin/tslint" --version
5.1.0
✨  Done in 0.44s.
```
- __TypeScript version__:
```
yarn run tsc -- --version
yarn run v0.22.0
$ "/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/node_modules/.bin/tsc" --version
Version 2.3.0
✨  Done in 0.29s.
```
- __Running TSLint via__: (pick one) CLI / Node.js API / VSCode / grunt-tslint / Atom / Visual Studio / etc
```
$ "/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/node_modules/.bin/tslint" -c tslint.json -p tsconfig.json --type-check
```
#### TypeScript code being linted

```ts
const { key: saveErrorKey, error: saveError } = (action as PayloadAction<KeyedError>).payload;
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "align": [
      true,
      "parameters",
      "arguments",
      "statements"
    ],
    "ban": false,
    "class-name": true,
    "comment-format": [
      true,
      "check-space"
    ],
    "curly": true,
    "eofline": true,
    "forin": true,
    "indent": [
      true,
      "spaces"
    ],
    "interface-name": false,
    "jsdoc-format": true,
    "label-position": true,
    "max-line-length": [
      false,
      140
    ],
    "member-ordering": [
      true,
      "public-before-private",
      "static-before-instance",
      "variables-before-functions"
    ],
    "no-any": false,
    "no-arg": true,
    "no-bitwise": true,
    "no-console": [
      true,
      "debug",
      "info",
      "time",
      "timeEnd",
      "trace"
    ],
    "no-consecutive-blank-lines": true,
    "no-construct": true,
    "no-parameter-properties": false,
    "no-debugger": true,
    "no-duplicate-variable": true,
    "no-shadowed-variable": true,
    "no-empty": false,
    "no-eval": true,
    "no-require-imports": false,
    "no-string-literal": true,
    "no-switch-case-fall-through": true,
    "trailing-comma": [true, {
        "multiline": "always",
        "singleline": "never"
    }],
    "no-trailing-whitespace": true,
    "no-unused-expression": true,
    "no-use-before-declare": true,
    "no-var-keyword": true,
    "no-var-requires": true,
    "one-line": [
      true,
      "check-catch",
      "check-else",
      "check-open-brace",
      "check-whitespace"
    ],
    "quotemark": [
      true,
      "double"
    ],
    "radix": true,
    "semicolon": true,
    "switch-default": true,
    "triple-equals": [
      true,
      "allow-null-check"
    ],
    "typedef": [
      false,
      "call-signature",
      "parameter",
      "property-declaration"
    ],
    "typedef-whitespace": [
      true,
      {
        "call-signature": "nospace",
        "index-signature": "nospace",
        "parameter": "nospace",
        "property-declaration": "nospace",
        "variable-declaration": "nospace"
      }
    ],
    "variable-name": [
      true,
      "allow-leading-underscore"
    ],
    "whitespace": [
      true,
      "check-branch",
      "check-decl",
      "check-operator",
      "check-separator",
      "check-type"
    ]
  }
}

```

#### Actual behavior
```
ERROR: /Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/src/redux/uiData.ts[92, 15]: variable 'key' used before declaration
ERROR: /Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/src/redux/uiData.ts[92, 34]: variable 'error' used before declaration
ERROR: /Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/src/redux/uiData.ts[109, 15]: variable 'key' used before declaration
ERROR: /Users/tamird/src/go/src/github.com/cockroachdb/cockroach/pkg/ui/src/redux/uiData.ts[109, 34]: variable 'error' used before declaration
```

#### Expected behavior
Clean.

**Comment by errorx666** [comment]
I just ran into this as well.

**Comment by sberan** [comment]
I have a minimal reproduction:

```
export function foo ({ param: p }: { param: any }) {
}
```

**Comment by sberan** [comment]
Any aliasing of the destructured param seems to trigger the issue.

**Comment by kalahari** [comment]
Is `"no-use-before-declare": false` the best work-around? The [documentation](https://palantir.github.io/tslint/rules/no-use-before-declare/) suggest it is only useful for `var` keyword variables anyway.

**Comment by abdullahsari** [comment]
This issue is still present, happened to me just now. Good stuff!

**Comment by piotros** [comment]
Problem still occurs.

**Comment by zheeeng** [comment]
Still exists

**Comment by lselbeck** [comment]
Without typescript I have no errors:
`const { email: user1Email, name: user1Name } = user1`

With typescript, I get the false positive:
`const { email: user1Email, name: user1Name }: { email: string, name: string } = user1`

**Comment by Krumpet** [comment]
Error still occurs, the 'variables' used before declaration in `const { label: l, shortcut: s }` are `label` and `shortcut`.

**Comment by kyro38** [comment]
Same problem !

**Comment by adidahiya** [comment]
This rule is essentially deprecated... check out the documentation https://palantir.github.io/tslint/rules/no-use-before-declare/
